File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/BottomTabBar/BottomTabBarModel.kt
// Pure, framework-free model + projection + diagnostics for the BottomTabBar shared surface — the native
// analogue of everything the web component derives before it returns its <nav> tree
// (web/src/components/layout/BottomTabBar.tsx). No Compose, no Android, no HTTP: every declaration here is
// unit-tested off-device in the :app:testReleaseUnitTest gate, keeping the composable a thin render layer.
//
// What the web source actually is (and therefore the COMPLETE branch set this surface reproduces): a fixed
// five-item mobile tab bar — Dashboard → Drives → Charging → Battery → Map — over `useLocation()`. The ONLY
// per-render decision the web makes is which single tab is active, computed per tab as:
//
//     isActive = tab.path === '/'
//       ? location.pathname === '/'                                  // the root tab matches ONLY exactly
//       : location.pathname === tab.path                             // an exact match, or…
//         || location.pathname.startsWith(tab.path + '/')            // …a descendant of the tab's section
//
// [BottomTabBarProjection.isActive] reproduces that algorithm verbatim — including the deliberate "+ '/'" that
// stops a sibling section from lighting the tab (e.g. `/charging-curve` must NOT activate the `/charging` tab,
// while `/charging/123` must). The five canonical destinations are taken from the shared
// [io.teslasync.android.navigation.RouteTable.bottomBar] set (which itself "mirrors the web BottomTabBar
// top-5"), so the native bar can never drift from the route table; [BottomTab.entries] is asserted equal to it
// in the projection test.
//
// Why the generic data-surface states (loading / error / stale / offline) are intentionally absent here: this
// surface fetches nothing. It is navigation chrome bound to the router state-holder (the `useLocation`
// analogue, P1/S8) whose only input is the already-resolved current path — always present, never a network
// lifecycle. Modelling loading/error/stale/offline would invent behaviour the web spec does not have (honesty
// covenant: no scope narrowing, no silent drift), exactly as the sibling router-driven RouteAnnouncer surface
// documents. The surface's REAL, fully-reproduced state set is: one-of-five tab active, a descendant of a
// section active, and no-tab-active (the current route lives outside the five sections) — each reduced here
// and asserted in the off-device projection test, and rendered (never hidden) by the composable.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/BottomTabBar — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName`
// is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.bottomtabbar

import io.teslasync.android.navigation.Destination
import io.teslasync.android.navigation.Destinations
import io.teslasync.android.navigation.RouteTable
import io.teslasync.shared.core.diagnostics.Logger

/**
 * Canonical registry metadata for this surface — the native mirror of the web component's contract. The
 * diagnostics slug is pinned here so the native and web surfaces stay in lockstep.
 */
object BottomTabBarRegistration {
    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "BottomTabBar"
}

/**
 * The five fixed tabs of the mobile bottom bar — the native port of the web `TABS` array, in the same order
 * (Dashboard → Drives → Charging → Battery → Map). Each tab binds to a canonical
 * [io.teslasync.android.navigation.Destination] by id, so the tab's [path] (the web `tab.path`), title key,
 * and icon are read from the single shared route table rather than re-declared. The id set is asserted equal
 * to [RouteTable.bottomBar] in the projection test, so this enum can never silently drift from the app's
 * canonical bottom-bar set.
 *
 * @property destinationId the stable [io.teslasync.android.navigation.Destination] id this tab targets.
 */
enum class BottomTab(
    val destinationId: String,
) {
    /** Web `{ path: '/', icon: Home, i18nKey: 'nav.dashboard' }`. */
    Dashboard("dashboard"),

    /** Web `{ path: '/drives', icon: Car, i18nKey: 'nav.drives' }`. */
    Drives("drives"),

    /** Web `{ path: '/charging', icon: BatteryCharging, i18nKey: 'nav.charging' }`. */
    Charging("charging"),

    /** Web `{ path: '/battery', icon: HeartPulse, i18nKey: 'nav.battery' }`. */
    Battery("batteryHealth"),

    /** Web `{ path: '/live', icon: MapPin, i18nKey: 'nav.liveMap' }`. */
    LiveMap("liveMap"),
    ;

    /** The canonical destination this tab navigates to (web `tab.path` target). */
    val destination: Destination get() = Destinations.require(destinationId)

    /** The tab's web path (web `tab.path`), e.g. `/`, `/drives`, `/charging`, `/battery`, `/live`. */
    val path: String get() = destination.webPath
}

/**
 * Localized labels the surface folds into its output. Built from `stringResource` at the render boundary
 * (tests pass a deterministic instance), keeping [BottomTabBarProjection] a pure, locale-stable function.
 * Every string resolves through the web component's exact i18n keys in the P1/S10 catalog: [navLabel] from
 * `nav.quickNav`, and one label per tab from `nav.dashboard` / `nav.drives` / `nav.charging` / `nav.battery` /
 * `nav.liveMap`.
 *
 * @property navLabel the bar's landmark label (web `aria-label={t('nav.quickNav', 'Quick navigation')}`).
 * @property dashboard the Dashboard tab label (web `nav.dashboard`).
 * @property drives the Drives tab label (web `nav.drives`).
 * @property charging the Charging tab label (web `nav.charging`).
 * @property battery the Battery tab label (web `nav.battery`).
 * @property liveMap the Map tab label (web `nav.liveMap`).
 */
data class BottomTabBarStrings(
    val navLabel: String,
    val dashboard: String,
    val drives: String,
    val charging: String,
    val battery: String,
    val liveMap: String,
) {
    /** The localized label for [tab] (web `t(tab.i18nKey, tab.fallback)`). */
    fun label(tab: BottomTab): String =
        when (tab) {
            BottomTab.Dashboard -> dashboard
            BottomTab.Drives -> drives
            BottomTab.Charging -> charging
            BottomTab.Battery -> battery
            BottomTab.LiveMap -> liveMap
        }
}

/**
 * One rendered tab in the bar — a [tab], its already-localized [label], and whether it is the [active]
 * destination for the current route. Pure data so the projection is unit-tested without a UI host; the
 * composable maps it onto a Material 3 `NavigationBarItem` (icon + label + selected state).
 */
data class BottomTabItem(
    val tab: BottomTab,
    val label: String,
    val active: Boolean,
) {
    /** The canonical destination a tap navigates to (web `<PrefetchLink to={tab.path} />`). */
    val destination: Destination get() = tab.destination
}

/**
 * The immutable, render-ready projection the composable draws — the bar's landmark [navLabel] plus the five
 * [items] with their active flags. Pure data so [BottomTabBarProjection] is unit-tested off-device.
 */
data class BottomTabBarDisplay(
    val navLabel: String,
    val items: List<BottomTabItem>,
) {
    /** The single active tab, or `null` when the current route lives outside all five sections. */
    val activeTab: BottomTab? get() = items.firstOrNull(BottomTabItem::active)?.tab
}

/**
 * Pure active-tab logic + projection for the BottomTabBar surface — the native port of the web component's
 * per-tab `isActive` derivation. Framework-free so the whole active-state contract is covered by the JVM unit
 * gate without a Compose host.
 */
object BottomTabBarProjection {
    /** The root path the Dashboard tab matches only exactly (web `tab.path === '/'`). */
    const val ROOT_PATH: String = "/"

    /**
     * Whether the tab at [tabPath] is active for [currentPath] — a verbatim port of the web `isActive`:
     * the root tab (`/`) matches ONLY an exact root path; every other tab matches its own path exactly OR any
     * descendant of it (`startsWith(tab.path + '/')`). The trailing slash is deliberate: it keeps a sibling
     * section from lighting the tab (`/charging-curve` does not activate `/charging`, while `/charging/123`
     * does). [currentPath] is normalized first so a stray trailing slash or query string can never flip the
     * result.
     */
    fun isActive(
        currentPath: String,
        tabPath: String,
    ): Boolean {
        val path = RouteTable.normalize(currentPath)
        return if (tabPath == ROOT_PATH) {
            path == ROOT_PATH
        } else {
            path == tabPath || path.startsWith("$tabPath/")
        }
    }

    /**
     * Folds the current route's [currentPath] (the `useLocation().pathname` analogue) together with the
     * localized [strings] into the render-ready [BottomTabBarDisplay]: the bar's landmark label and the five
     * tabs with each one's active flag resolved by [isActive]. The current path is normalized once so the
     * five per-tab checks share a single canonical input.
     */
    fun project(
        currentPath: String,
        strings: BottomTabBarStrings,
    ): BottomTabBarDisplay {
        val normalized = RouteTable.normalize(currentPath)
        val items =
            BottomTab.entries.map { tab ->
                BottomTabItem(
                    tab = tab,
                    label = strings.label(tab),
                    active = isActive(normalized, tab.path),
                )
            }
        return BottomTabBarDisplay(navLabel = strings.navLabel, items = items)
    }
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the route
 * path nor the active tab — so a diagnostics line can never leak which screen a user is on.
 */
object BottomTabBarDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = BottomTabBarRegistration.SLUG

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}

File: apps/windows/TeslaSync.App/shared-surfaces/BottomTabBar/BottomTabBar.Model.cs
using System.Collections.Generic;
using System.Globalization;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata + i18n keys for the BottomTabBar surface — the native mirror of the web
/// <c>BottomTabBar</c> (web/src/components/layout/BottomTabBar.tsx), the mobile bottom navigation rail
/// pinned under the viewport on small breakpoints (<c>lg:hidden</c>). The web source is presentational:
/// its destinations are the frozen module-level <c>TABS</c> constant and its only hooks are
/// <c>useLocation</c> (the active-path source) and <c>useTranslation</c> (i18n), so it performs no data
/// fetch — and therefore renders no loading / error / stale / offline chrome to reproduce. This metadata
/// carries the diagnostics slug the surface registers under and the navigation-landmark i18n
/// key/fallback the web source passes to <c>t('nav.quickNav', 'Quick navigation')</c>; the per-tab keys
/// live verbatim on each <see cref="BottomTab"/> in the catalogue. Each key carries the
/// <c>translation.</c> catalog prefix the WinUI resource bridge expects and resolves against the English
/// fallback headlessly. UI-free so it is asserted without a XAML host.
/// </summary>
public static class BottomTabBarRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "BottomTabBar";

    /// <summary>i18n key for the navigation landmark accessible name (web <c>t('nav.quickNav', 'Quick navigation')</c>).</summary>
    public const string QuickNavKey = "translation.nav.quickNav";

    /// <summary>English fallback for <see cref="QuickNavKey"/> (web second arg, verbatim).</summary>
    public const string QuickNavFallback = "Quick navigation";

    /// <summary>The root route whose tab is active only on an exact match (web <c>tab.path === '/'</c>).</summary>
    public const string RootPath = "/";

    /// <summary>
    /// i18n key for the defensive empty-state copy. The web bar has no empty state (its <c>TABS</c> are a
    /// fixed non-empty constant), but the native surface still renders a friendly message instead of a blank
    /// bar if it is ever handed an empty catalogue — never a blank box. Reuses the shared catalog key
    /// <c>common.noData</c> so no new string is introduced.
    /// </summary>
    public const string EmptyMessageKey = "translation.common.noData";

    /// <summary>English fallback for <see cref="EmptyMessageKey"/>.</summary>
    public const string EmptyMessageFallback = "No data available";
}

/// <summary>
/// One render-ready tab — the pure projection of a <see cref="BottomTab"/> against the current path and
/// the i18n facade. <see cref="Label"/> is the localized text shown under the glyph AND the tab's
/// accessible name (the web source sets both the <c>&lt;span&gt;</c> text and the link's
/// <c>aria-label</c> from the same <c>t(tab.i18nKey, tab.fallback)</c> value). <see cref="IsActive"/> is
/// the web <c>isActive</c> branch — it drives the accent foreground, the icon glow, the bottom accent
/// pill and the <c>aria-current="page"</c> state.
/// </summary>
/// <param name="Path">Route the tab navigates to (web <c>tab.path</c>); the click intent + active key.</param>
/// <param name="Glyph">Segoe Fluent Icons code point shown above the label.</param>
/// <param name="Label">Localized label + accessible name (web <c>t(tab.i18nKey, tab.fallback)</c>).</param>
/// <param name="IsActive">Whether the current path resolves to this tab (web <c>isActive</c>).</param>
public sealed record BottomTabDisplay(string Path, string Glyph, string Label, bool IsActive);

/// <summary>
/// The render-ready projection of the whole bar — the localized navigation-landmark name plus the
/// ordered, already-resolved tabs. <see cref="IsEmpty"/> is true only when the catalogue is empty (the
/// fixed web <c>TABS</c> never is, but the surface still renders a friendly empty state instead of a
/// blank bar so a region never collapses silently); the web source has no loading / error / stale /
/// offline projection because it performs no fetch.
/// </summary>
/// <param name="NavAutomationName">Localized navigation-landmark accessible name (web <c>aria-label</c> on the <c>&lt;nav&gt;</c>).</param>
/// <param name="Tabs">The ordered, projected tabs (web <c>TABS.map(...)</c>).</param>
/// <param name="IsEmpty">Whether there are no tabs to render (drives the empty state).</param>
/// <param name="EmptyMessage">Localized friendly empty-state copy shown when <paramref name="IsEmpty"/> is true.</param>
public sealed record BottomTabBarDisplay(
    string NavAutomationName,
    IReadOnlyList<BottomTabDisplay> Tabs,
    bool IsEmpty,
    string EmptyMessage);

/// <summary>
/// The pure, UI-thread-free projection behind the BottomTabBar view — the native unification of the web
/// source's <c>isActive</c> computation and its <c>TABS.map</c> render loop. It owns the active-path rule
/// verbatim and resolves every label through the i18n facade, so the view-model and tests share one
/// render contract and the view only data-binds the result. Stateless; safe to call from any thread.
/// </summary>
public static class BottomTabBarProjection
{
    /// <summary>
    /// Whether <paramref name="tabPath"/> is the active tab for <paramref name="currentPath"/> — the web
    /// <c>isActive</c> rule verbatim: the root route ("/") matches only an exact path, while every other
    /// tab matches an exact path OR a descendant path (<c>pathname.startsWith(tab.path + '/')</c>) so
    /// nested routes such as <c>/drives/42</c> keep the Drives tab lit. Ordinal throughout — route paths
    /// are culture-invariant identifiers, never display text.
    /// </summary>
    public static bool IsActive(string? currentPath, string tabPath)
    {
        ArgumentNullException.ThrowIfNull(tabPath);
        string path = Normalize(currentPath);

        if (string.Equals(tabPath, BottomTabBarRegistration.RootPath, StringComparison.Ordinal))
        {
            return string.Equals(path, BottomTabBarRegistration.RootPath, StringComparison.Ordinal);
        }

        return string.Equals(path, tabPath, StringComparison.Ordinal)
            || path.StartsWith(tabPath + "/", StringComparison.Ordinal);
    }

    /// <summary>
    /// Project the catalogue against the current path and the i18n facade into the render-ready
    /// <see cref="BottomTabBarDisplay"/> — the native <c>TABS.map</c>. Resolves the navigation-landmark
    /// name and each tab's label/active state once so the view performs no logic. Recomputed on demand
    /// (the catalogue is five entries) so it always reflects the latest path and language.
    /// </summary>
    /// <param name="tabs">The tab catalogue (web <c>TABS</c>); typically <see cref="BottomTabBarCatalog.Default"/>.</param>
    /// <param name="currentPath">The live route path (web <c>location.pathname</c>); null/empty normalizes to "/".</param>
    /// <param name="localizer">The i18n facade every label resolves through (web <c>useTranslation</c>, P1/S10).</param>
    public static BottomTabBarDisplay Project(
        IReadOnlyList<BottomTab> tabs,
        string? currentPath,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(tabs);
        ArgumentNullException.ThrowIfNull(localizer);

        string path = Normalize(currentPath);
        string navName = localizer.GetString(
            BottomTabBarRegistration.QuickNavKey, BottomTabBarRegistration.QuickNavFallback);
        string emptyMessage = localizer.GetString(
            BottomTabBarRegistration.EmptyMessageKey, BottomTabBarRegistration.EmptyMessageFallback);

        var projected = new List<BottomTabDisplay>(tabs.Count);
        foreach (BottomTab tab in tabs)
        {
            string label = localizer.GetString(tab.TitleKey, tab.Fallback);
            bool active = IsActive(path, tab.Path);
            projected.Add(new BottomTabDisplay(tab.Path, tab.Glyph, label, active));
        }

        return new BottomTabBarDisplay(navName, projected, projected.Count == 0, emptyMessage);
    }

    /// <summary>Normalize a route path the way the web router guarantees: a null / empty path is the root "/".</summary>
    public static string Normalize(string? path) =>
        string.IsNullOrEmpty(path) ? BottomTabBarRegistration.RootPath : path;
}

/// <summary>
/// PII-safe diagnostics for the BottomTabBar surface (P1/S11 diagnostics contract). A tab bar's
/// destinations are route paths and labels that can hint at a user's feature usage, so the collector
/// records ONLY the operational <see cref="RecordViewOpened"/> signal with the surface slug — never a
/// route or a label. Thread-safe; mirrors the shipped surfaces' collectors.
/// </summary>
public sealed class BottomTabBarDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public BottomTabBarDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=BottomTabBar</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(
            string.Create(
                CultureInfo.InvariantCulture,
                $"view.opened slug={BottomTabBarRegistration.Slug}"));
    }
}
